pkg/multiboot/multiboot.go

```diff
@@ -0,0 +1,185 @@
+// Copyright 2018 the u-root Authors. All rights reserved
+// Use of this source code is governed by a BSD-style
+// license that can be found in the LICENSE file.
+
+// Package multiboot implements basic primitives
+// to load multiboot kernels as defined in
+// https://www.gnu.org/software/grub/manual/multiboot/multiboot.html.
+package multiboot
+
+import (
+	"fmt"
+	"io"
+	"log"
+	"os"
+	"unsafe"
+
+	"github.com/u-root/u-root/pkg/kexec"
+)
+
+const (
+	memmapRoot = "/sys/firmware/memmap/"
+	bootloader = "u-root kexec"
+)
+
+// Multiboot defines parameters for working with multiboot kernels.
+type Multiboot struct {
+	file string
+
+	cmdLine    string
+	bootloader string
+
+	header Header
+	// Segments are memory segments to be loaded by kexec.
+	Segments segments
+	ranges   memoryRanges
```

Comment:
I'm wondering if there's a better way to store this information. I don't have a clear picture in mind yet, but I'm just a little internally bothered by the fact that we have two different ways of representing physical memory here -- segments to be passed to kexec, and reserved memory ranges -- and the allocation routine has to deal with checking both of them.

Here's what I'm thinking -- the physical memory allocation bits, the parsing sysfs memmap stuff, and the segment representation will be common to all kexec image formats (bzimage, multiboot, EFI).

Can we turn memoryRanges into kexec.PhysicalMemory or something like that, and build it as a separate allocator?

I'm thinking there would be another rangeType 'kexecSegment'

Also, some of the range stuff in memoryRange could be done by the existing kexec.Range type.

```go
package kexec

type PhysicalMemory []AddressRange

// type memoryRange turns into
type AddressRange struct {
  Mem Range
  Type RangeType
  // some way to store the Buf Range thing for RangeType = KexecSegment? dunno
}

func (pm PhysicalMemory) ToSegments() Segments {
  for _, ar := range pm.Filter(TypeKexecSegment) {
     // ...
  }
  return kexec segments
}
```

the multiboot package can have its own function to convert to mmap:
```go
package multiboot

func MemoryMap(pm *kexec.PhysicalMemory) ... {
  for _, ar := range pm {
    if ar.Type != TypeKexecSegment {
       // ... add to thing
     }
   }
}
```

there are some missing pieces in my head for this but... think about it...